### ISSUE TITLE
Conditional annotation

### ISSUE THREAD

**Issue by twinkarma** [GateNLP/gate-teamware#164]
Suggested by Xingi

Depending on the selected option in a label, additional labels can be shown for further annotation, creating a hierarchy of annotations.

**Comment by ianroberts** [comment]
@iknoorjobs has also raised a possible requirement for changing which and how many input widgets are rendered based on the value of fields from the document, e.g. given a document:

```json
{
  "query": "This is a search query",
  "results": [
    {"docId": "8009", "text":"Result document 1"},
    {"docId": "3693", "text": "Result document 2"}
  ]
}
```

we want to gather "relevant" = yes/no/unknown, and "confidence" = 1-10 _for each result_, i.e.

> ### Query
>
> This is a search query
>
> ### Result 8009
>
> > #### Text
> > Result document 1
> >
> > #### Relevant?
> > [ ] yes   [ ] unsure    [ ] no
> > #### Confidence
> > [selector 0-10]
>
> ### Result 3693
>
> > #### Text
> > Result document 2
> >
> > #### Relevant?
> > [ ] yes   [ ] unsure    [ ] no
> > #### Confidence
> > [selector 0-10]

producing output along the lines of

```json
{
  "results": [
    {"relevant": "yes", "confidence": 8},
    {"relevant": "no", "confidence": 3}
  ]
}
```

The question is how flexible does this need to become?  Can we specify a reasonable set of scenarios that are nicely implementable without making the JSON too complex, or are we basically talking about putting a Turing-complete programming language into the config parser...

**Comment by twinkarma** [comment]
What I think we can do is provide a "dynamic" widget that allows a user to define per-document widget configuration. This, combined with allowing hierarchical widgets inside the "dynamic" widget as well should cover most cases? 

It will be on the project manager to generate a suitable widget configuration for each document.

**Comment by ianroberts** [comment]
Yeah, I was thinking something similar.  Two new widget types should be sufficient, one for "if" and one for "foreach".

### Conditional execution

```json
{
  "type": "conditional",
  "if": "{expression}",
  "children":[
    // list of widgets
  ]
}
```

We could use something like https://www.npmjs.com/package/jexl for the expression parser, and give it a context including the current document and the current state of the other annotation fields (to cover the original use case of things like "if user ticks 'other' then show field prompting for free text").

### Iteration

```json
{
  "name": "docs",
  "type": "repeat",
  "in": "property.path",
  "var": "current",
  "children":[
    // list of widgets
  ]
}
```

This would extract `property.path` from the document and iterate over it:

- if the value at `property.path` is an array, then the annotation result from this widget will be an array of objects with the annotation results from `children`
- if the value at `property.path` is an object, then the annotation result from this widget will be an object with the same keys, each mapped to an object with the annotation results from `children`

In both cases the child widgets would be passed a copy of the current document data with an extra property named by `var` (so `current` in the above example) referring to the current iteration item.  So you could nest these repeats using `"in": "current.something", "var":"somethingElse"`

#### Full example

**Document**
```json
{
  "query": "This is a search query",
  "results": [
    {"docId": "8009", "text":"Result document 1"},
    {"docId": "3693", "text": "Result document 2"}
  ]
}
```

**Config**
```json
[
  {
    "type":"html",
    "name": "query",
    "title": "Search query",
    "text": "{{query}}"
  },
  {
    "type": "repeat",
    "name": "results",
    "in": "results",
    "var": "res",
    "children": [
      {
        "type": "html",
        "name": "doc",
        "text": "<h3>{{res.docId}}</h3><p>{{res.text}}</p>"
      },
      {
        "type": "radio",
        "name": "relevant",
        "title": "Relevant?",
        "options": [
           {"value":"yes", "label": "Yes"},
           {"value":"maybe", "label": "Maybe"},
           {"value":"no", "label": "No"}
        ]
      },
      {
        "type": "selector",
        "name": "confidence",
        "title": "Confidence",
        "options": ["0", "1", "2", ... ]
      }
    ]
  }
]
```

In `<h3>{{res.docId}}</h3><p>{{res.text}}</p>`, the `res` refers to the current iteration item.  We could also add a `"status"` variable to expose either the array index (if iterating an array) or current item key (if iterating an object).

**Comment by ianroberts** [comment]
Or it might make more sense to allow _any_ widget to have the "if" and "repeat" options, and just add one extra widget type for grouping:

```json
{
  "type":"group",
  "name":"mygroup",
  "title": "Title for this group",
  "children": [ ... ]
}
```

which would nest the results from all its "children" components under a "mygroup" key in the output.

For conditionals we just allow any widget to have an `"if": "expression"` property (if you want a conditional group then you just use "if" with a "group" widget), and for iteration we allow any widget to have a `"foreach": "property.path", "var": "variableName", "status": "indexOrKeyVariable"` following the same rules I suggested above, so if:

```json
{
  "name": "example",
  "type":"text",
}
```

produces

```json
{ "example": "user-input-text" }
```

then

```json
{
  "name": "example",
  "type": "text",
  "foreach": "results",
  "var": "res"
}
```

would produce

```json
{
  "example": ["user text for doc 0", "user text for doc 1", ...]
}
```

And a `foreach` _group_ widget would produce an array of objects as in my previous comment.

For this to work nicely we'd probably also have to allow the "title" and "description" fields to be mustache templates in the same way as the HTML display widget so that the `var` and/or `status` could be substituted into each widget description.

**Comment by ianroberts** [comment]
Question - should the `foreach` be limited to a document-side property path or is there a case to make it the full expression language same as `if` with access to the rest of the live widget values?  This would open up really dynamic cases like a multi-select set of checkboxes where each time you click one a new text field pops up asking "why did you select {{value}}?"

**Comment by twinkarma** [comment]
I think it makes sense to allow every widget to have the `foreach` property  and we might as well use the full expression language that's already implemented with `if`. I'd imagine it should not be much more effort to implement @ianroberts?

I'd suggest having a separate `group` widget which will also support how components are laid out inside it allowing you to lay out multiple columns or allow custom css scoped just inside the widget?

e.g. for a multi-column layout:
```
{
   "type":"group",
   "orientation":"horizontal", <-- provides a quick way to define directionality of components, defaults to horizontal?
   "class":"", <-- Standard css, probably want this to override the orientation flag completely
   "style":"", <-- if either class or style is set
   "children":[
      {
         "type":"group",
         "children":[
            {
               "type":"html", ...
            }
         ]
      }
   ]
}
```

Also we might as well allow `class` and `style` property in all widgets if we're going to be implementing it in the `group` widget

**Comment by ianroberts** [comment]
> I think it makes sense to allow every widget to have the `foreach` property and we might as well use the full expression language that's already implemented with `if`. I'd imagine it should not be much more effort to implement @ianroberts?

Indeed, it shouldn't be too hard - we might want to add some sort of `map` and `filter` syntax to the expression parser along the same lines as the `all` and `any` predicates but in principle it sounds fine.

The tricky part once we start introducing `foreach` is working out which _other_ parts of the config need to become templates or expressions, and what context needs to be available to those.  For example, within a `foreach` you might want to interpolate the current iteration's value into the `title` or `description` of the repeated field (or in the case of a repeated group, _one of_ the fields within the repeating group, and then groups can be nested so you have to track a whole lexical scope at each stage).  And if you're `foreach`-ing over a more complex structure like an array of objects then a pure Mustache template won't necessarily be enough (e.g. `"description": "Why did you choose the {{cur.label}} option?"` where you need to access a property of the current iteration value).
